### ISSUE TITLE
Group filters with opacity 

### ISSUE THREAD

**Issue by sriranga19** [CyberAgent/android-gpuimage#260]
Hi,
It is very useful. Thanks for providing such a beautiful code.

I walk through your code and it seems some thing is missing in group filter( Group filters with opacity ) i.e., If i want to apply sepia(GPUImageSepiaFilter.java) with 60% of opacity and an "GPUImageLookupFilter.java" with 70%. Is it possible to apply above case using group filter?

**Comment by dockusan** [comment]
Did you want sepia photo 60% alpha blend with lookup photo 70% alpha?

**Comment by dockusan** [comment]
I don't understand what you do? But I think you should decrease intensity in filter.
i.e: groupFilter.addFilter(new Sepia(0.6f);
     groupFilter.addFilter(new Lookup(0.7f);

**Comment by sriranga19** [comment]
Thanks for your response. As you ask above what i need exactly "Did you want sepia photo 60% alpha blend with lookup photo 70% alpha?" but in place of sepia i have another LUT.i.e, I have two LUTs .I want to apply first LUT filter with 40% opacity and second LUT with 50% opacity using group filter.

My code snippet is given below.

filter = applyEffectFilter(MainActivity.this,lutbitmap,values[i]);
         filters.add(filter);
public static GPUImageFilter applyEffectFilter(Context con,Bitmap lutBitmap,float value)
 {
  GPUImageLookupFilter filter = new GPUImageLookupFilter(value);
  filter.setBitmap(lutBitmap);
  
  return filter;
  
 }

where values[]={0.4,0.5}.

**Comment by dockusan** [comment]
Did you resolve this problem?

**Comment by sriranga19** [comment]
NO

**Comment by sriranga19** [comment]
After getting list i am doing below one. Please correct me if i am doing wrong. 

GPUImageFilter filter= new GPUImageFilterGroup(filters);

        
        GPUImage gpuimage = new GPUImage(this);
        gpuimage.setImage(original);
        gpuimage.setFilter(filter);         image.setImageBitmap(gpuimage.getBitmapWithFilterApplied(original));

**Comment by dockusan** [comment]
Sorry. I want to know what is LUT here

**Comment by sriranga19** [comment]
Its a image like "lookup_amatorka.png" in sample which is used in GPUlookup filter.

**Comment by dockusan** [comment]
![filter](https://cloud.githubusercontent.com/assets/10556072/13740541/e0f653d2-ea03-11e5-83b5-8b3b210b54e8.png)

**Comment by dockusan** [comment]
Oh. I think you make right. you must set gpuimage.setFilter(filters) ("filters" : group filter contain 2 LookupFilter)

**Comment by tyson512** [comment]
Hi dockusan, thanks for your reply i am also facing the same problem what sriranga19 facing so i tried the above process for two lookup filters but result was not correct(i.e getting black image).

here is the code for two lookup filters :

 `GPUImage gpuimage = new GPUImage(this);
        gpuimage.setImage(original);
        
        
        GPUImageFilterGroup group = new GPUImageFilterGroup();
        
       Bitmap lutbitmap = BitmapFactory.decodeResource(getResources(), R.drawable.poolparty);
        Bitmap result = applyEffectBitmap(MainActivity.this, original,lutbitmap);
        
        GPUImageLookupFilter lookup = new GPUImageLookupFilter(0.0f);
        lookup.setBitmap(result);
        group.addFilter(lookup);
        
        gpuimage.setFilter(group);
        
        Bitmap lutbitmap1 = BitmapFactory.decodeResource(getResources(), R.drawable.cube_terra);
        Bitmap result1 = applyEffectBitmap(MainActivity.this, gpuimage.getBitmapWithFilterApplied(),lutbitmap1);
        
        GPUImageLookupFilter lookup1 = new GPUImageLookupFilter(1.0f);
        lookup1.setBitmap(result1);
        group.destroy(); // here i didn't find clear() in group so i called destory() 
        group.addFilter(lookup1);
        gpuimage.setFilter(group);
        
        
        imageView.setImageBitmap(gpuimage.getBitmapWithFilterApplied());` 

so please let me know the correct process..

**Comment by tyson512** [comment]
Hi dockusan, thanks for your reply i am also facing the same problem what sriranga19 facing so i tried the above process for two lookup filters but result was not correct(i.e getting black image).

here is the code for two lookup filters :

`GPUImage gpuimage = new GPUImage(this);
gpuimage.setImage(original); // original bitmap

    GPUImageFilterGroup group = new GPUImageFilterGroup();

   Bitmap lutbitmap = BitmapFactory.decodeResource(getResources(), R.drawable.lut1);
    Bitmap result = applyEffectBitmap(MainActivity.this, original,lutbitmap);

    GPUImageLookupFilter lookup = new GPUImageLookupFilter(0.0f);
    lookup.setBitmap(result);
    group.addFilter(lookup);

    gpuimage.setFilter(group);

    Bitmap lutbitmap1 = BitmapFactory.decodeResource(getResources(), R.drawable.lut2);
    Bitmap result1 = applyEffectBitmap(MainActivity.this, gpuimage.getBitmapWithFilterApplied(),lutbitmap1);

    GPUImageLookupFilter lookup1 = new GPUImageLookupFilter(1.0f);
    lookup1.setBitmap(result1);
    group.destroy(); // here i didn't find clear() in group so i called destory() 
    group.addFilter(lookup1);
    gpuimage.setFilter(group);


    imageView.setImageBitmap(gpuimage.getBitmapWithFilterApplied());`
so please let me know the correct process..

**Comment by dockusan** [comment]
group.destroy(); // here i didn't find clear() in group so i called destory() 
Please change to:  group.getFilters().clear();

**Comment by tyson512** [comment]
Thanks for reply @dockusan . black image issue was cleared but still i'm facing issue with intensity value. Getting same output image for intensity values 0.0f and 1.0f i.e 
GPUImageLookupFilter lookup1 = new GPUImageLookupFilter(0.0f);   and 
GPUImageLookupFilter lookup1 = new GPUImageLookupFilter(1.0f);

Hope 0.0f need to produce image without filter effect and 1.0f needs to produce image with filter effect but i am getting filter applied image in both cases.

**Comment by dockusan** [comment]
You should check if(value ==0) -> setFilter( new GPUImageFilter());
This filter will make filter image to normal.

**Comment by tyson512** [comment]
@dockusan yes i did it, it worked but still i have an issue can you please suggest me the solution for if i pass values between 0.1 to 0.9.

**Comment by dockusan** [comment]
Oh. I got it. You must set opacity for lutbitmap.
```
Bitmap lutbitmap1 = BitmapFactory.decodeResource(getResources(), R.drawable.lut2);
lutbitmap1 = setAlpha(value) (0-1 ~ 0 -255) 
```
Then apply LookupFilter without set value. 
[Link to make bitmap opacity](http://stackoverflow.com/questions/5118894/how-to-change-a-bitmaps-opacity)

**Comment by tyson512** [comment]
i think it may work but i think i'm doing some thing wrong in the process of applying two lookup filters because i'm getting last filter with alpha only as a result. 

here is the code :

` GPUImage gpuimage = new GPUImage(this);
        gpuimage.setImage(original); // original bitmap

        GPUImageFilterGroup group = new GPUImageFilterGroup();
        Bitmap lutbitmap = BitmapFactory.decodeResource(getResources(), R.drawable.lut1);

        GPUImageLookupFilter lookup = new GPUImageLookupFilter(0.8f);
        lookup.setBitmap(lutbitmap);
        group.addFilter(lookup);

        gpuimage.setFilter(group);
        
        image1.setImageBitmap(gpuimage.getBitmapWithFilterApplied());

        Bitmap lutbitmap1 = BitmapFactory.decodeResource(getResources(), R.drawable.cube_terra);
        Bitmap newbitmap = Bitmap.createBitmap(lutbitmap1.getWidth(),lutbitmap1.getHeight(), lutbitmap1.getConfig());
        
        Canvas can = new Canvas(newbitmap);
        Paint alphaPaint = new Paint();
        alphaPaint.setAlpha(200);
        can.drawBitmap(lutbitmap1, 0, 0, alphaPaint);
        
        try {
        	
        	File file = new File(Environment.getExternalStorageDirectory()+"/sample/bitmap.png");
			FileOutputStream fos = new FileOutputStream(file);
			newbitmap.compress(Bitmap.CompressFormat.PNG, 100, fos);
		} catch (FileNotFoundException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
        
        
        GPUImageLookupFilter lookup1 = new GPUImageLookupFilter();
        
        lookup1.setBitmap(newbitmap);
        Bitmap bit =gpuimage.getBitmapWithFilterApplied();
        gpuimage.deleteImage();
        gpuimage.setImage(bit);
        group.getFilters().clear(); 
        group.addFilter(lookup1);
        
        gpuimage.setFilter(group);
        imageView.setImageBitmap(gpuimage.getBitmapWithFilterApplied());`   

In this i just applied alpha for second filter only. 

can you please check it and guide me where i am doing wrong?
